src/esp/gfx_batch/Renderer.cpp

```diff
@@ -522,17 +553,47 @@ bool Renderer::addFile(const Cr::Containers::StringView filename,
         return {};
       }
 
-      const auto& flatMaterial = material->as<Mn::Trade::FlatMaterialData>();
-      materials[i] = Mn::Shaders::PhongMaterialUniform{}.setAmbientColor(
-          flatMaterial.color());
-
-      /* Untextured materials get the first reserved texture (a white pixel) */
-      if (!flatMaterial.hasTexture())
-        materialTextureTransformations[i] = {0, 0, {}};
-      else
-        materialTextureTransformations[i] = {
-            flatMaterial.texture() + textureOffset, flatMaterial.textureLayer(),
-            flatMaterial.textureMatrix()};
+      /* Set up flat shading either if the material is marked as flat or if
+         we have no lights enabled */
+      if (material->types() & Mn::Trade::MaterialType::Flat ||
+          !state_->maxLightCount) {
+        const auto& flatMaterial = material->as<Mn::Trade::FlatMaterialData>();
+        materials[i] = Mn::Shaders::PhongMaterialUniform{}
+                           .setAmbientColor(flatMaterial.color())
+                           /* Diffuse is zero so lights (if enabled) have no
+                              effect on this material  */
+                           .setDiffuseColor(0x00000000_rgbaf);
+
+        /* Untextured materials get the first reserved texture (a white
+           pixel) */
+        if (!flatMaterial.hasTexture())
+          materialTextureTransformations[i] = {0, 0, {}};
+        else
+          materialTextureTransformations[i] = {
+              flatMaterial.texture() + textureOffset,
+              flatMaterial.textureLayer(), flatMaterial.textureMatrix()};
+      } else {
+        const auto& phongMaterial =
+            material->as<Mn::Trade::PhongMaterialData>();
+        materials[i] = Mn::Shaders::PhongMaterialUniform{}
+                           .setAmbientColor(phongMaterial.diffuseColor() * 0.1f)
```

Comment:
Ug, what is this 0.1? Are we hard-coding the amount of ambient light?

Reply:
Yes -- see [this Slack post for details](https://cvmlp.slack.com/archives/G0131KVLBLL/p1671575565719199?thread_ts=1671497345.493199&cid=G0131KVLBLL).

I wanted to achieve an acceptable look until there's a proper PBR pipeline where one doesn't need to hand-tune this kind of thing anymore. Same for specular highlights. The materials are all encoded as PBR, which means there's no info about "classic" specular highlights or ambient anymore.

Reply:
How about re-using `LightSetup.h getAmbientLightColor`? If we matched that value, would it be easier to get visual parity with the classic renderer?

Reply:
I could look into that, yup -- I remember you spent a considerable amount of time on the PBR -> Phong conversion, so it would be a shame to ignore it.

How should I deal with the original requirement that `gfx_batch` stays completely isolated from the rest of Habitat? (I was considering putting some PBR -> Phong conversion directly into Magnum, but that time would probably be better spent on actually making PBR work.)

But, re visual parity -- the recent Floorplanner discussion made me think I should maybe try something from scratch instead of trying to match how it looks in the classic renderer :sweat_smile:

Reply:
>the original requirement that gfx_batch stays completely isolated from the rest of Habitat?

Yep. So this would be some caller like replayer.cpp passing `LightSetup.h getAmbientLightColor` to the batch renderer, maybe via a new public method or the Configuration struct.

>the recent Floorplanner discussion made me think I should maybe try something from scratch instead of trying to match how it looks in the classic renderer
I don't have the full context here, but I think it'll be valuable going forward to be able to swap between the classic and batch renderer, with Mikael working to reach parity.

Reply:
Oh, wait, this is about the *light* ambient color, not the per-material ambient color, I missed that.

So, just to clarify we're on the same page -- the aspect of diffuse / base color used partially for ambient is fine in this case, and you're only concerned about the factor being hardcoded, right? Not about the fact that the same texture is used for both.

Having it configurable from the light makes sense, but it'll be a global factor (i.e., there's no way to have different lights with different ambient factors affecting different meshes, the ambient factors will get all summed up and used as a single value). But that should be fine, I think?

Reply:
>you're only concerned about the factor being hardcoded, right?
Yes.

>Not about the fact that the same texture is used for both.
No problem here. I expect this.

Somewhat related, I know we've disagreed in this past a bit on this nuance of the Phong lighting model. My preference is for a material's ambient color to generally match its diffuse color, and for that material ambient color to get multiplied with an ambient light color at runtime (every frame), such that you could hypothetically vary the color/amount of ambient light in the scene every frame. Whereas I know you've spoke in the past of preferring to pre-multiply these two and basically bake them into the material's ambient color. I'm too lazy to track down that Slack thread from a year ago, but more specifically, you said that many 3D assets in the wild already do this pre-baking by convention. Anyway, we don't have to debate this now or change the current code in this respect.

Reply:
>Having it configurable from the light makes sense, but it'll be a global factor 
Exactly, a global factor, not per light. I think of ambient light color as sorta like the cosmic background radiation. It's a constant amount of light shining in all directions, everywhere in the scene. A lazy proxy for e.g. env-map-based light.